builtin/logical/ssh/path_config_ca.go

```diff
@@ -265,22 +281,89 @@ func (b *backend) pathConfigCAUpdate(ctx context.Context, req *logical.Request,
 	return nil, nil
 }
 
-func generateSSHKeyPair(randomSource io.Reader) (string, string, error) {
+func generateSSHKeyPair(randomSource io.Reader, keyType string, keyBits int) (string, string, error) {
 	if randomSource == nil {
 		randomSource = rand.Reader
 	}
-	privateSeed, err := rsa.GenerateKey(randomSource, 4096)
-	if err != nil {
-		return "", "", err
-	}
 
-	privateBlock := &pem.Block{
-		Type:    "RSA PRIVATE KEY",
-		Headers: nil,
-		Bytes:   x509.MarshalPKCS1PrivateKey(privateSeed),
+	var publicKey interface{}
+	var privateBlock *pem.Block
+
+	switch keyType {
+	case ssh.KeyAlgoRSA:
+		if keyBits == 0 {
+			keyBits = 4096
+		}
+
+		if keyBits < 2048 {
+			return "", "", fmt.Errorf("refusing to generate weak %v key: %v bits < 2048 bits", keyType, keyBits)
+		}
+
+		privateSeed, err := rsa.GenerateKey(randomSource, keyBits)
+		if err != nil {
+			return "", "", err
+		}
+
+		privateBlock = &pem.Block{
+			Type:    "RSA PRIVATE KEY",
+			Headers: nil,
+			Bytes:   x509.MarshalPKCS1PrivateKey(privateSeed),
+		}
+
+		publicKey = &privateSeed.PublicKey
```

Comment:
I'd rather us use the Public() method on the crypto algos as they behave a little differently across the various implementations :(

```suggestion
		publicKey = privateSeed.Public()
```